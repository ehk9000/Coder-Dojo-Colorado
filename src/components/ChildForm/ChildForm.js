import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import requests from '../../utils/requests/requests';
import * as actions from '../../actions';

export class ChildForm extends Component {
  state = {
    name: '',
    username: '',
    password: '',
    password2: '',
    dob: '',
    success: false
  }

  handleChange = (e) => {
    const { name, value } = e.target;
    
    this.setState({
      [name]: value
    });
  }

  handleSubmit = (e) => {
    e.preventDefault();

    this.props.setError('');

    const error1 = this.checkAllFields();
    const error2 = this.checkPasswords();
    
    if (!error1 && !error2) {
      console.log('there were no errors')
    } 
  }

  checkAllFields() {
    let error = false;
    const { 
      name, 
      username, 
      password, 
      password2, 
      dob 
    } = this.state;
    const { setError } = this.props;
    const fields = [name, username, password, password2, dob]; 

    fields.forEach((field) => {
      if (!field) {
        setError('All fields must be filled.');
        error = true;
      } 
    });

    return error;
  }

  checkPasswords() {
    let error = false;
    const { password, password2 } = this.state;
    const { setError } = this.props;

    if (password !== password2) {
      setError('Passwords must match.');

      this.setState({
        password: '',
        password2: ''
      });

      error = true;
    } 

    return error;
  }

  registerStudent = async () => {
    const { 
      name, 
      username, 
      password, 
      dob 
    } = this.state;
    const student = {
      name,
      username,
      password,
      birthdate: dob
    };
    const createResult = await requests.createStudent(student);

    if (createResult) {
      this.setState({ success: true });
    }
  }

  render() {
    const { password, password2, success } = this.state;
    const { error, loading } = this.props;

    if (success) return <Redirect to="/" />;

    return (
      <form 
        className="ChildForm"
        onSubmit={this.handleSubmit}
      >
        <h2>Register a Student</h2>
        <label htmlFor="child-name-input">
          Student's Full Name
          <input
            type="text"
            id="child-name-input"
            name="name"
            onChange={this.handleChange}
          />
        </label>
        <label htmlFor="child-dob">
          Student's Date of Birth
          <input 
            type="text"
            id="child-dob"
            name="dob"
            onChange={this.handleChange}
          />
        </label>
        <label htmlFor="child-username-input">
          Student's Username
          <input
            type="text"
            id="child-username-input"
            name="username"
            onChange={this.handleChange}
          />
        </label>
        <label htmlFor="child-password-input">
          Student's Password
          <input
            type="password"
            id="child-password-input"
            name="password"
            value={password}
            onChange={this.handleChange}
          />
        </label>
        <label htmlFor="child-password2-input">
          Re-Enter Student's Password
          <input
            type="password"
            id="child-password2-input"
            name="password2"
            value={password2}
            onChange={this.handleChange}
          />
        </label>
        <p className="error-msg">{error && error}</p>
        <button type="submit">
          {loading ? 'Please wait...' : 'Submit'}
        </button>
      </form>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  setError: error => dispatch(actions.setError(error))
});

export const mapStateToProps = state => ({
  error: state.error,
  loading: state.isFetching
});

export default connect(mapStateToProps, mapDispatchToProps)(ChildForm);

ChildForm.propTypes = {
  setError: PropTypes.func,
  error: PropTypes.string,
  loading: PropTypes.bool
};

ChildForm.defaultProps = {
  setError: () => {},
  error: '',
  loading: false
};
