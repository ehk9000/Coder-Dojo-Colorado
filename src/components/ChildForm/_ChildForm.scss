.ChildForm {
  @include form-style;

  h2 {
    margin: 0 0 15px 0;
    text-align: center;
  }

  label {
    @include flex-column;
  }

  input {
    @include form-input-style;
  }

  .error-msg {
    @include form-error-msg;
  }

  button {
    @include form-submit($footer-color);
  }
}
