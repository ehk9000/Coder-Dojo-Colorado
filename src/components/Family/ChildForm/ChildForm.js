import React from 'react';

function ChildForm() {
  return (
    <div>

    </div>
  );
}

export default ChildForm;
