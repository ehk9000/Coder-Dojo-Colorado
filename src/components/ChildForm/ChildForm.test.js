import React from 'react';
import { shallow } from 'enzyme';
import { ChildForm } from './ChildForm';

describe('ChildForm', () => {
  let wrapper;
  let setError;

  beforeEach(() => {
    setError = jest.fn();
    wrapper = shallow(
      <ChildForm 
        setError={setError}
        error={''}
        loading={false}
      />
    );
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('handleChange', () => {
    it('should update state on change of any input field', () => {
      const mockEvent = {
        target: {
          value: 'abc',
          name: 'name'
        }
      };

      wrapper.instance().handleChange(mockEvent);

      expect(wrapper.state('name')).toEqual('abc');
    });
  });

  describe('handleSubmit', () => {
    const mockEvent = {
        preventDefault: jest.fn()
      };

    it('should prevent default on submit', () => {
      wrapper.instance().handleSubmit(mockEvent);

      expect(mockEvent.preventDefault).toHaveBeenCalled();
    });
  });

  describe('registerStudent', () => {
    const student = {
      name: 'Jude Bacher',
      username: 'jude',
      password: 'password',
      birthdate: '01/23/18'
    }

    it.skip('should create a student in the db', async () => {
      const result = await requests.createStudent(student);
    });
  });
  
  describe('checkAllFields', () => {
    it.skip('should set an error if a field is empty', () => {
      wrapper.setState({
        name: '',
        username: 'judebacher',
        password: 'password',
        password: 'password',
        dob: '1/23/18'
      });

      wrapper.instance().checkAllFields();

      expect(wrapper.setError).toHaveBeenCalledWith('All fields must be filled.');
    });
  });

  describe('checkPasswords', () => {
    it.skip('should set an error if the passwords do not match', () => {
      wrapper.setState({
        name: 'Jude Bacher',
        username: 'judebacher',
        password: 'password',
        password: 'passwordd',
        dob: '1/23/18'
      });

      wrapper.instance().checkPasswords();

      expect(wrapper.setError).toHaveBeenCalledWith('Passwords must match.');
    });
  });

  describe('mapStateToProps', () => {

  });

  describe('mapDispatchToProps', () => {

  });
});
