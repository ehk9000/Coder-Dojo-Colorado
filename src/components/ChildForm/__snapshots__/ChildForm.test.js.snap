// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChildForm should match snapshot 1`] = `
<form
  className="ChildForm"
  onSubmit={[Function]}
>
  <h2>
    Register a Student
  </h2>
  <label
    htmlFor="child-name-input"
  >
    Student's Full Name
    <input
      id="child-name-input"
      name="name"
      onChange={[Function]}
      type="text"
    />
  </label>
  <label
    htmlFor="child-dob"
  >
    Student's Date of Birth
    <input
      id="child-dob"
      name="dob"
      onChange={[Function]}
      type="text"
    />
  </label>
  <label
    htmlFor="child-username-input"
  >
    Student's Username
    <input
      id="child-username-input"
      name="username"
      onChange={[Function]}
      type="text"
    />
  </label>
  <label
    htmlFor="child-password-input"
  >
    Student's Password
    <input
      id="child-password-input"
      name="password"
      onChange={[Function]}
      type="password"
      value=""
    />
  </label>
  <label
    htmlFor="child-password2-input"
  >
    Re-Enter Student's Password
    <input
      id="child-password2-input"
      name="password2"
      onChange={[Function]}
      type="password"
      value=""
    />
  </label>
  <p
    className="error-msg"
  />
  <button
    type="submit"
  >
    Submit
  </button>
</form>
`;
